bionic/deriver.py

```diff
@@ -459,29 +461,38 @@ def _process_entry(self, entry, task_key_logger):
         # Initialize the task state before attempting to complete it.
         state.initialize(self._bootstrap, self._flow_instance_uuid)
 
-        # self._bootstrap.process_executor is None when parallel processing is
-        # disabled or when entities are bootstrapped.
         if (
-            not state.provider.attrs.should_persist()
+            # Bootstrap entity.
+            self._bootstrap is None
+            # Serial processing.
             or self._bootstrap.process_executor is None
+            # Non-serializable entity that needs to be returned.
+            or (
+                not state.provider.attrs.should_persist()
+                and entry.state.task.keys[0] in self._requested_task_keys
+            )
         ):
-            # NOTE 1: This makes non-persisted entities compute here as well
-            # as in executor pool. We need to keep track of what needs to be
-            # computed in main process vs subprocess when entity is not persisted.
-            # NOTE 2: Right now, non-persisted entities include simple lookup values
+            # NOTE: Right now, non-persisted entities include simple lookup values
             # which we should not be really sending using IPC. We should read/write
             # a tmp file for this instead to use protocol for serialization instead of
             # using cloudpickle.
             state.complete(task_key_logger)
             self._mark_entry_completed(entry)
-        else:
-            # NOTE 1: Logging support for multiple processes not done yet.
+
+        # Process serializable entity in parallel.
+        elif state.provider.attrs.should_persist():
+            # NOTE: Logging support for multiple processes not done yet.
```

Comment:
Would it make sense for these `NOTE`s to be `TODO`s?

Reply:
Yeah, makes sense to me.